Userland/Utilities/sed.cpp

```diff
@@ -0,0 +1,559 @@
+        for (auto& command : script.commands())
+            command.enable_for(pattern_space.string_view(), input.line_number(), is_last_line);
+
+        // Go, go, go!
+        CycleDecision cycle_decision = CycleDecision::None;
+        for (auto& command : script.commands()) {
+            if (!command.is_enabled())
+                continue;
+            auto command_cycle_decision = apply(command, pattern_space, hold_space, input, suppress_default_output);
+            if (command_cycle_decision == CycleDecision::Next || command_cycle_decision == CycleDecision::Quit) {
+                cycle_decision = command_cycle_decision;
+                break;
+            }
+        }
+
+        if (cycle_decision == CycleDecision::Next) {
+            continue;
+        } else if (cycle_decision == CycleDecision::Quit) {
+            break;
+        }
+        if (!suppress_default_output)
+            write_pattern_space(stdout, pattern_space);
+        pattern_space.clear();
+    }
+
+    return true;
+}
+
+int main(int argc, char* argv[])
+{
+    bool suppress_default_output = false;
+    Core::ArgsParser arg_parser;
+    Script script;
+    Vector<const char*> pos_args;
+    arg_parser.set_general_help("The Stream EDitor");
+    arg_parser.add_option(suppress_default_output, "suppress default output", nullptr, 'n');
+    arg_parser.add_option(Core::ArgsParser::Option {
+        .requires_argument = true,
+        .help_string = "A file containing script commands",
+        .short_name = 'f',
+        .value_name = "script-file",
+        .accept_value = [&script](const char* script_file) {
+            auto file = Core::File::open(script_file, Core::OpenMode::ReadOnly);
+            if (file.is_error()) {
+                perror("Failed to open script file");
+                return false;
+            }
+            return script.add_script_part(file.value()->read_all());
+        },
+    });
+    arg_parser.add_option(Core::ArgsParser::Option {
+        .requires_argument = true,
+        .help_string = "A script of commands",
+        .short_name = 'e',
+        .value_name = "script",
+        .accept_value = [&script](const char* script_argument) {
+            return script.add_script_part(script_argument);
+        },
+    });
+    arg_parser.add_positional_argument(pos_args, "script and/or file", "...", Core::ArgsParser::Required::No);
+    if (!arg_parser.parse(argc, argv)) {
+        VERIFY_NOT_REACHED();
+    }
+
+    if (script.commands().is_empty()) {
+        if (pos_args.is_empty()) {
+            warnln("No script specified, aborting");
+            return 1;
+        }
+        if (!script.add_script_part(pos_args[0])) {
+            return 1;
+        }
+        pos_args.remove(0);
+    }
+    Vector<InputFile> inputs;
+    for (const auto& filename : pos_args) {
+        auto file = Core::File::open(filename, Core::OpenMode::ReadOnly);
+        if (file.is_error()) {
+            perror("Cannot open input file for reading");
```

Comment:
Same note about `perror` and `File::open()` here.

Reply:
Fixed.